File: Beginner/Valid Parenthesis/ValidParenthesis.cpp
// Using string
// T: O(n) 
// S: O(1) 

#include <string>
#include <iostream>
using namespace std;

class Solution {
public:
    bool isValid(std::string s) {
        int i = -1;
        for (auto& ch: s) {
            if (ch == '(' || ch == '{' || ch == '[')
                s[++i] = ch;
            else {
                if (i>=0 && (ch==')' && s[i]=='(' || ch=='}' && s[i]=='{' || ch==']' && s[i]=='['))
                    i--;
                else
                    return false;
            }
        }
        return i==-1;
    }

};

int main() {
    Solution s;
    std::string expr = "{()}[]";

    if (s.isValid(expr))
        std::cout << "Valid";
    else
        std::cout << "Invalid";

    return 0;
}

File: Beginner/Valid Parenthesis/ValidParenthesis.py
# Using Stack
# T: O(n) 
# S: O(n) for  the stack
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        lookup = {')': '(', '}': '{', ']': '['}

        for p in s:
            if p in lookup.values():
                stack.append(p)
            elif stack and lookup[p] == stack[-1]:
                stack.pop()
            else:
                return False
        return stack == []

s = Solution()
print(f"This is: {s.isValid('()')}")        # True
print(f"This is: {s.isValid('()[]{}')}")    # True
print(f"This is: {s.isValid('(]')}")        # False